### PR TITLE
Matrix rework

### PR DESCRIPTION
**Pull request by rjeli**

### CODE REVIEW

**Comment by rjeli** [comment]
Summary:

#### Only one (immutable) `Matrix` trait.
The backing matrix only needs to provide a row iterator and the rest is implemented by default. However, a wrapper should implement `get` and `row_slice`, as those are inefficient if relying on the row iterator default impl. `Matrix<T>` now has `Send + Sync` bounds, so parallel row iterators work seamlessly, and it's fine because we're always immutable.

`Matrix` provides `row_slice(&self) -> impl AsRef<[T]>`, so if you really need random access, you can get it, but this possibly allocates a Vec and returns that (for instance, an Extension -> Base flattening matrix view won't be able to get the underlying slice). I don't think we'll hit this path very often, though. `AsRef<[T]>` is necessary so it can return either a slice or vec. (`Borrow<[T]>` would also work).

Any general compute-heavy field routines should go on the `Matrix` trait. The default implementations can take advantage of `par_rows()`. Right now it's just `columnwise_dot_product`. It should be fully parallelized for even, e.g. bit-reversed matrix views, but can be further specialized for DenseMatrix in the future to do exotic packing.

Generally, I have gone to painstaking lengths to ensure any method can be overridden by the underlying matrix implementation.

#### `DenseMatrix<T, Storage: Borrow<[T]>>` to unify `RowMajorMatrix{View{Mut}}`

As `Matrix<T>` is always immutable, if you want to mutate the matrix, you need to use `RowMajorMatrix`. I have rewritten it so it's generic over any backing slice, deduplicating some code.

Btw, the main reason `Matrix` is immutable is to avoid `Send + Sync` bounds proliferation. I don't foresee needing to support it in the future, but if we do, there is a possible efficient API for row-wise mutation, for example, for matrices that are simply stacking or permuting a dense matrix.

#### Simplify butterflies

Idk, this isn't that important, it just looks nicer this way, plus got rid of unnecessary packing prefix

#### `MMCS` commits to and holds onto any `Matrix`

Instead of only taking `RowMajorMatrix`, `Mmcs` now takes a generic `Matrix<T>` and will return it as a GAT: `Mmcs::ProverData<M>`. Then, when you ask for a reference back (`get_matrices`), it will return the original matrix. This doesn't immediately seem useful but it turns out to simplify a lot of APIs and allow for zero-copy. For instance, it avoids copying in `ExtensionMmcs`, and will avoid further copies for bit-reversed or circle-permuted matrices. In the near future, we can put methods on `Matrix` to do the vertical packing for hashing even faster.

#### Add `SharedExt` to `maybe_rayon`

Some methods like `fold` and `reduce` have different signatures on `Iterator` and `ParallelIterator`, so added another trait to unify the APIs for methods we want to use.

**Comment by rjeli** [comment]
TLDR:
```
pub trait Matrix<T: Send + Sync>: Send + Sync {
    // default implementation
    fn columnwise_dot_product<EF>(&self, v: &[EF]) -> Vec<EF>
    where
        T: Field,
        EF: ExtensionField<T>,
    {
        self.par_rows().zip(v).par_fold_reduce(
            || vec![EF::zero(); self.width()],
            |mut acc, (row, &scale)| {
                izip!(&mut acc, row).for_each(|(a, x)| *a += scale * x);
                acc
            },
            |mut acc_l, acc_r| {
                izip!(&mut acc_l, acc_r).for_each(|(l, r)| *l += r);
                acc_l
            },
        )
    }
}
```
magically goes fast

**Comment by rjeli** [comment]
Btw, I went back and forth a few times between having `Row` be `IntoIterator` or `Iterator`. `IntoIterator` is conceptually slightly nicer, but if you ever want bounds on `Row`, e.g. `Send + Sync`, it gets messy having to bound both `Row` and `<Row as IntoIterator>::IntoIter`, and often the messy bounds propagate downstream.

**Comment by rjeli** [comment]
While several of the benefits are merely nice-to-haves, the primary motivation for this PR is allow for greater control over memory usage. Previously we would have to work on `RowMajorMatrix` when we wanted to do something specialized or low-level, and be forced to convert to that representation. Now we can choose whether to convert to dense or operate directly on a wrapper type, while still maintaining decent (and specializable) performance in the latter case.

File: matrix/src/lib.rs [hunk 1]
```diff
@@ -52,25 +44,38 @@ impl Display for Dimensions {
...
     }
 
-    fn row_vec(&self, r: usize) -> Vec<T> {
-        self.row(r).into_iter().collect()
+    fn par_rows(&self) -> impl IndexedParallelIterator<Item = Self::Row<'_>> {
+        (0..self.height()).into_par_iter().map(move |r| self.row(r))
+    }
+
+    // Opaque return type implicitly captures &'_ self
+    fn row_slice(&self, r: usize) -> impl Deref<Target = [T]> {
```
Comment (dlubarov) [inline matrix/src/lib.rs:77]: Interesting, hadn't thought of doing it this way... I think it makes AIRs a tad less nice but maybe worth it for the more unified API

Comment (rjeli) [reply matrix/src/lib.rs:77]: Yeah, I changed from `AsRef<[T]>` to `Deref<Target = [T]>` so you can just do `local[0]` instead of `local.as_ref()[0]`. Still sometimes need to fiddle with it though, like having to do `(*local).borrow()` for the column casting

File: matrix/src/lib.rs [hunk 2]
```diff
@@ -92,66 +97,56 @@ pub trait MatrixRows<T>: Matrix<T> {
...
+        (0..self.width()).map(move |c| P::from_fn(|i| self.get(r + i, c)))
+    }
+
     fn vertically_strided(self, stride: usize, offset: usize) -> VerticallyStridedMatrixView<Self>
     where
         Self: Sized,
     {
+        VerticallyStridedPerm::new_view(self, stride, offset)
+        /*
         VerticallyStridedMatrixView {
```
Comment (dlubarov) [inline matrix/src/lib.rs:127]: remove?

Comment (rjeli) [reply matrix/src/lib.rs]: oops ty

File: air/src/two_row_matrix.rs [hunk 3]
```diff
@@ -1,7 +1,6 @@
-use core::iter::Cloned;
-use core::slice;
+use core::{iter, slice};
 
-use p3_matrix::{Matrix, MatrixRowSlices, MatrixRows};
+use p3_matrix::Matrix;
 
 #[derive(Copy, Clone, Debug)]
 pub struct TwoRowMatrixView<'a, T> {
```
Comment (dlubarov) [inline air/src/two_row_matrix.rs:6]: Should we impl `row_slice`, to avoid the default which allocates? (Or maybe not after optimizations)

File: commit/src/mmcs.rs [hunk 4]
```diff
@@ -19,37 +19,47 @@ use serde::Serialize;
 /// The `DirectMmcs` sub-trait represents an MMS which can be directly constructed from a set of
```
Comment (dlubarov) [inline commit/src/mmcs.rs:19]: out of date comment

File: dft/src/radix_2_bowers.rs [hunk 5]
```diff
@@ -77,80 +75,54 @@ impl<F: TwoAdicField> TwoAdicSubgroupDft<F> for Radix2Bowers {
...
 
     let root = F::two_adic_generator(log_h);
-    let mut twiddles: Vec<F> = root.powers().take(h / 2).collect();
+    let mut twiddles: Vec<_> = root.powers().take(h / 2).map(DifButterfly).collect();
     reverse_slice_index_bits(&mut twiddles);
 
     let log_h = log2_strict_usize(mat.height());
     for log_half_block_size in 0..log_h {
-        bowers_g_layer(mat, log_half_block_size, &twiddles);
+        // bowers_g_layer(mat, log_half_block_size, &twiddles);
```
Comment (dlubarov) [inline dft/src/radix_2_bowers.rs:88]: nice refactor, seems we can remove this

File: dft/src/radix_2_bowers.rs [hunk 6]
```diff
@@ -77,80 +75,54 @@ impl<F: TwoAdicField> TwoAdicSubgroupDft<F> for Radix2Bowers {
...
-) {
-    let half_block_size = 1 << log_half_block_size;
-    let width = mat.width();
-    par_chunks_bowers(mat, width, half_block_size, twiddles, dit_butterfly_on_rows)
-}
-
-fn par_chunks_bowers<F: Field, Fun>(
-    mat: &mut RowMajorMatrixViewMut<'_, F>,
-    width: usize,
+fn par_chunks_bowers<F: Field, B: Butterfly<F>>(
```
Comment (dlubarov) [inline dft/src/radix_2_bowers.rs:109]: nit: we could call this something like `butterfly_layer` now that it's generic to the point there's no Bowers logic (which I guess it was already, just more obvious after your refactor)

File: interpolation/src/lib.rs [hunk 7]
```diff
@@ -45,11 +43,13 @@ where
         .collect();
     let diff_invs = batch_multiplicative_inverse(&diffs);
 
-    let mut sum = vec![EF::zero(); width];
-    for (i, (subgroup_i, diff_inv)) in g.powers().zip(diff_invs).enumerate() {
-        let row = coset_evals.row(i).into_iter();
-        add_scaled_base_slice_in_place(&mut sum, row, diff_inv * subgroup_i);
-    }
+    let col_scale: Vec<_> = g
```
Comment (dlubarov) [inline interpolation/src/lib.rs:46]: nit: maybe add a TODO to vectorize this part also (shouldn't be too hard since we have `powers_packed`)

File: matrix/src/permuted.rs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+use p3_field::PackedValue;
+
+use crate::dense::RowMajorMatrix;
+use crate::Matrix;
+
+/// A RowPermutation remaps row indices, and can change the height.
+pub trait RowPermutation: Send + Sync {
+    fn height(&self) -> usize;
+    fn permute_row_index(&self, r: usize) -> usize;
```
Comment (dlubarov) [inline matrix/src/permuted.rs:11]: Maybe some of the language should be tweaked since this isn't really a permutation (e.g. when striding). Maybe `map_`, `convert_` or `transform_` in place of `permute_`?

File: matrix/src/extension.rs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    fn height(&self) -> usize {
+        self.0.height()
+    }
+
+    type Row<'a> = FlatIter<F, Inner::Row<'a>>
+    where
+        Self: 'a;
+
+    fn row(&self, r: usize) -> Self::Row<'_> {
+        FlatIter {
```
Comment (dlubarov) [inline matrix/src/extension.rs:39]: I wonder if we could later change `row` to return `impl Iterator<...>`, so implementations like this could leverage `flat_map`?

File: matrix/src/lib.rs [hunk 10]
```diff
@@ -92,66 +97,49 @@ pub trait MatrixRows<T>: Matrix<T> {
...
+    where
+        Self: Sized,
+    {
+        VerticallyStridedPerm::new_view(self, stride, offset)
     }
-}
 
-pub trait MatrixRowChunksMut<T: Send>: MatrixRowSlicesMut<T> {
-    type RowChunkMut<'a>: MatrixRowSlicesMut<T> + Send
+    fn columnwise_dot_product<EF>(&self, v: &[EF]) -> Vec<EF>
```
Comment (dlubarov) [inline matrix/src/lib.rs:128]: nit: let's add a comment, might not be totally clear from the name what it does. I think one way we could describe it is "premultiply this matrix by the given vector"

**Review by dlubarov — APPROVED** [review]
LGTM

File: matrix/src/extension.rs (hunk 9 above)
Comment (rjeli) [reply matrix/src/extension.rs:39]: So I tried to hack it this way for a while, but it eventually gets annoying not being able to name the Row type, especially for Send + Sync bounds. You end up with nested impl traits and HRTBs for things like par_rows that percolate downstream

File: air/src/two_row_matrix.rs (hunk 3 above)
Comment (rjeli) [reply air/src/two_row_matrix.rs:6]: Good catch
